src/main/drivers/sbus_output.c

```diff
@@ -0,0 +1,174 @@
+    return 0;
+}
+
+// Note: this returns 1000x normalized value (-1000 - 1000).
+static float sbusOutGetValueMixer(uint8_t channel)
+{
+    if (channel < MIXER_OUTPUT_COUNT)
+        return 1000.0f * mixerGetOutput(channel);
+    return 0;
+}
+
+static float sbusOutGetServo(uint8_t channel)
+{
+    if (channel < MAX_SUPPORTED_SERVOS)
+        return getServoOutput(channel);
+    return 0;
+}
+
+// Note: this returns 0 - 1000 range (or -1000 - 1000 for bidirectional
+// motor)
+static float sbusOutGetMotor(uint8_t channel)
+{
+    if (channel < MAX_SUPPORTED_MOTORS)
+        return getMotorOutput(channel);
+    return 0;
+}
+
+STATIC_UNIT_TESTED float sbusOutGetChannelValue(uint8_t channel)
+{
+    const sbusOutSourceType_e source_type =
+        sbusOutConfig()->sourceType[channel];
+    const uint8_t source_index = sbusOutConfig()->sourceIndex[channel];
+    switch (source_type) {
+    case SBUS_OUT_SOURCE_NONE:
+        return 0;
+    case SBUS_OUT_SOURCE_RX:
+        return sbusOutGetRX(source_index);
+    case SBUS_OUT_SOURCE_MIXER:
+        return sbusOutGetValueMixer(source_index);
+    case SBUS_OUT_SOURCE_SERVO:
+        return sbusOutGetServo(source_index);
+    case SBUS_OUT_SOURCE_MOTOR:
+        return sbusOutGetMotor(source_index);
+    }
+    return 0;
+}
+
+STATIC_UNIT_TESTED uint16_t sbusOutConvertToSbus(uint8_t channel, float pwm)
+{
+    const int16_t low  = sbusOutConfig()->sourceRangeLow[channel];
+    const int16_t high = sbusOutConfig()->sourceRangeHigh[channel];
+
+    // round and bound values
+    if (channel >= 16) {
+        const float value = scaleRangef(pwm, low, high, 0, 1);
+        return constrain(nearbyintf(value), 0, 1);
+    }
+    const float value = scaleRangef(pwm, low, high, 192, 1792);
+    return constrain(nearbyintf(value), 0, (1 << 11) - 1);
+}
+
+void sbusOutUpdate(timeUs_t currentTimeUs)
+{
+    UNUSED(currentTimeUs);
+    if (!sbusOutPort)
+        return;
+
+    // Check TX Buff is free
+    if (serialTxBytesFree(sbusOutPort) <= sizeof(sbusOutFrame_t))
+        return;
+
+    // Start sending.
+    sbusOutFrame_t frame;
+    uint16_t channels[SBUS_OUT_CHANNELS];
+    for (int ch = 0; ch < SBUS_OUT_CHANNELS; ch++) {
+        float value = sbusOutGetChannelValue(ch);
+        channels[ch] = sbusOutConvertToSbus(ch, value);
+    }
+    sbusOutPrepareSbusFrame(&frame, channels);
+
+    // serial output
+    serialWriteBuf(sbusOutPort, (const uint8_t *)&frame, sizeof(frame));
+}
+
+bool sbusOutIsEnabled() 
+{ 
+    return sbusOutPort != NULL;
+}
+
+void sbusOutInit()
+{
+    const serialPortConfig_t *portConfig =
+        findSerialPortConfig(FUNCTION_SBUS_OUT);
+
+    if (!portConfig) {
+        sbusOutPort = NULL;
+        return;
+    }
+
+    sbusOutPort = openSerialPort(
+        portConfig->identifier, FUNCTION_SBUS_OUT, NULL, NULL, 100000, MODE_TX,
+        SERIAL_STOPBITS_2 | SERIAL_PARITY_EVEN | SERIAL_INVERTED |
+                SERIAL_BIDIR | sbusOutConfig()->pinSwap
```

Comment:
I just noticed this.
SERIAL_BIDIR with MODE_TX does not make much sense.
Can you explain your thinking?

Reply:
No idea what I was thinking. [b71de22](https://github.com/rotorflight/rotorflight-firmware/pull/190/commits/b71de22666552909d52485ad9dcffc485d85498f)